homeassistant/components/unifiprotect/media_source.py

```diff
@@ -0,0 +1,802 @@
+            return _bad_identifier_media(item.identifier, err)
+
+        event = data.api.bootstrap.events.get(parts[2])
+        if event is None:
+            try:
+                event = await data.api.get_event(parts[2])
+            except NvrError as err:
+                return _bad_identifier_media(item.identifier, err)
+            else:
+                # cache the event for later
+                data.api.bootstrap.events[event.id] = event
+
+        nvr = data.api.bootstrap.nvr
+        if thumbnail_only:
+            return PlayMedia(
+                async_generate_thumbnail_url(event.id, nvr.id), "image/jpeg"
+            )
+        return PlayMedia(async_generate_event_video_url(event), "video/mp4")
+
+    async def async_browse_media(self, item: MediaSourceItem) -> BrowseMediaSource:
+        """Return a browsable UniFi Protect media source.
+
+        Identifier formatters for UniFi Protect media sources are all in IDs from
+        the UniFi Protect instance since events may not always map 1:1 to a Home
+        Assistant device or entity. It also drasically speeds up resolution.
+
+        The UniFi Protect Media source is timebased for the events recorded by the NVR.
+        So its structure is a bit different then many other media players. All browsable
+        media is a video clip. The media source could be greatly cleaned up if/when the
+        frontend has filtering supporting.
+
+        * ... Each NVR Console (hidden if there is only one)
+            * All Cameras
+            * ... Camera X
+                * All Events
+                * ... Event Type X
+                    * Last 24 Hours -> Events
+                    * Last 7 Days -> Events
+                    * Last 30 Days -> Events
+                    * ... This Month - X
+                        * Whole Month -> Events
+                        * ... Day X -> Events
+
+        Accepted identifier formats:
+
+        * {nvr_id}
+            Root NVR source
+        * {nvr_id}:event:{event_id}
+            Specific Event for NVR
+        * {nvr_id}:eventthumb:{event_id}
+            Specific Event Thumbnail for NVR
+        * {nvr_id}:all|{camera_id}
+            Root Camera(s) source
+        * {nvr_id}:all|{camera_id}:all|{event_type}
+            Root Camera(s) Event Type(s) source
+        * {nvr_id}:all|{camera_id}:all|{event_type}:recent:{day_count}
+            Listing of all events in last {day_count}, sorted in reverse chronological order
+        * {nvr_id}:all|{camera_id}:all|{event_type}:browse:{year}:{month}
+            List of folders for each day in month + all events for month
+        * {nvr_id}:all|{camera_id}:all|{event_type}:browse:{year}:{month}:all|{day}
+            Listing of all events for give {day} + {month} + {year} combination in chronological order
+        """
+
+        if not item.identifier:
+            return await self._build_sources()
+
+        parts = item.identifier.split(":")
```

Comment:
Would a named tuple or something more flexible make sense here to avoid the magic numbers below?

Reply:
I am not sure I follow. The identifier has to be a string.

Reply:
I was hoping the data could be a bit more normalized and the string could be unpacked into something more structured like

```

@dataclass
class ItemIdentifer:

    nvr_id: str
    event: str | None
    event_id: str | None

...
```

Reply:
I refactored it a bit to cut down on that, but having a single class it not really possible as it branches too much.

Reply:
👍  Thats a lot cleaner.